arm: core: Initial Stack Guarding implementation

fvincenzo opened the pull request:
This pull request is a follow up of https://gerrit.zephyrproject.org/r/#/c/12499/

Commit:
arm: soc: nxp k6x: Add Initial support for NXP MPU

This patch adds initial MPU support to NXP K6x family.
The boot configuration prevents the following security issues:
* Prevent to read at an address that is reserved in the memory map.
* Prevent to write into the boot Flash/ROM.
* Prevent from running code located in SRAM.

This driver has been tested on FRDM-K64F.

Change-Id: I907168fff0c6028f1c665f1d3c224cbeec31be32
Signed-off-by: Vincenzo Frascino <vincenzo.frascino@linaro.org>

Commit:
arm: core: mpu: Add arm core MPU interface

This patch adds the arm core MPU interface, a common way to access the
pu functionalities by the arm zephyr kernel.

The interface can be divided in two parts:
- a core part that will be implemented by the arm_core_mpu driver and
  used directly by the kernel
- a driver part that will be implemented by the mpu drivers and used by
  the arm_core_mpu driver

Change-Id: I590bd284abc40d98b06fdf1efb5800903313aa00
Signed-off-by: Vincenzo Frascino <vincenzo.frascino@linaro.org>

Commit:
arm: core: mpu: Add core support to ARM MPU

This patch add arm core MPU support to ARM MPU driver.

Change-Id: I5a61da4615ae687bf42f1c9947e291ebfd2d2c1d
Signed-off-by: Vincenzo Frascino <vincenzo.frascino@linaro.org>

Commit:
arm: core: mpu: Add core MPU implementation

This patch adds the arm core MPU implementation.
This implementation currently supports the thread stack guard feature.

Change-Id: I8b3795ebaf1ebad38aaddc2ed2f05535ead2c09a
Signed-off-by: Vincenzo Frascino <vincenzo.frascino@linaro.org>

Commit:
arm: core: Integrate thread stack guard feature

This patch integrates the thread stack guard feature in the arm
Zephyr core.

Change-Id: I2022899cbc7a340be71cfaa52f79418292f93bae
Signed-off-by: Vincenzo Frascino <vincenzo.frascino@linaro.org>

Commit:
samples: synch: Add Thread Stack Guards conf

This patch adds the Thread Stack Guards configuration option to the
synchronization sample.

This is meant to be an easy way to prove that the MPU can be dynamically
programmed on the supported platforms.

Change-Id: I77fbe6af3303ff8a2904768abd4cc9797f03c092
Signed-off-by: Vincenzo Frascino <vincenzo.frascino@linaro.org>

Commit:
samples: Add mpu stack guard test

This patch adds the mpu_stack_guard_test to the available samples.
The purpose of this test is to exercize the thread stack guard enabled
via MPU on arm platforms.

Change-Id: I4665a20956d9e6d0dd4b5cc862e82040a53afafc
Signed-off-by: Vincenzo Frascino <vincenzo.frascino@linaro.org>

dbkinder reviewed (commented):
+1 docs look OK

MaureenHelm commented on arch/arm/soc/nxp_kinetis/k6x/Kconfig.soc:
```diff
@@ -96,6 +96,14 @@ config K64_FLASH_CLOCK_DIVIDER
+	default n
+	help
+	  Enable MPU
+
```
This should be one level higher, in arch/arm/soc/nxp_kinetis/Kconfig, because the sysmpu exists on other parts outside the k6x series.

MaureenHelm commented on arch/arm/soc/nxp_kinetis/k6x/nxp_mpu_regions.c:
```diff
@@ -0,0 +1,49 @@
+	 */
+	MPU_REGION_ENTRY("FLEXBUS_0",
+			 FLEXBUS_BASE_ADDRESS,
+			 0x1FFEFFFF,
```
Per my gerrit review, I'm not sure this address is correct. I think it should be 0x1BFF_FFFF?

MaureenHelm commented on arch/arm/core/cortex_m/mpu/nxp_mpu.c:
```diff
@@ -0,0 +1,128 @@
+static void _nxp_mpu_clock_cfg(void)
+{
+	/* Enable Clock */
+	SIM->SCGC7 |= SIM_SCGC7_MPU_MASK | SIM_SCGC7_FLEXBUS_MASK;
```
The SCGC registers are SoC-specific, so this isn't going to work across SoCs. Take a look at ext/hal/nxp/mcux/drivers/fsl_sysmpu.c in the SYSMPU_Init() function for a more portable example to enable the clock. You should be able to do this:
```
CLOCK_EnableClock(kCLOCK_Sysmpu0);
```

fvincenzo replied:
Ok, I will do. Thanks

MaureenHelm commented on boards/arm/frdm_k64f/frdm_k64f_defconfig:
```diff
@@ -10,3 +10,6 @@ CONFIG_GPIO=y
 CONFIG_HAS_DTS=y
+
+# Enable MPU
+CONFIG_HAS_SYSMPU=y
```
Do we really want to enable this by default? By disabling execution from SRAM, we'll break anything that uses the kinetis flash driver. For example, mcuboot.

fvincenzo replied:
I think that an application that has specific needs has to override this setting via its own config file. The default should be protecting against basic issues.

mbolivar replied:
I think we need to discuss how to approach applications that want to read and write flash in general. Right now, Carbon is already broken with mcuboot for MPU-related reasons (in addition to https://github.com/zephyrproject-rtos/zephyr/pull/122).

Out of tree, we are maintaining patches like this to keep our own builds working (aside from the applications patches):

https://github.com/Linaro/zephyr/commit/14af77b21ce54e10a0c62139c988271f48776b23
https://github.com/Linaro/zephyr/commit/0b2a39cedf5ab1ed4d90efa361b14de89f347e06

I think the MPU support patches that are coming in are awesome -- they've already helped us catch bugs, some of which are in upstream Zephyr. But it would be good to discuss in particular what the plans are for applications which need to read and write flash and access "other" areas of memory like the UID bytes or system memory on STM32.

fvincenzo replied:
First of all thanks for the info, I was not aware of it. This is definitely something that we need to target in the near future and I am going to send to you a separate email so we can formalize the requirements. 

Said that the patch you are referring to with your changes https://github.com/zephyrproject-rtos/zephyr/commit/5951e45580c198c11e8fac59544195123a46e3ad has been in RFC for nearly 2 months and I was hoping to have captured all the requirements in terms of exceptions but it seems something it is still missing :-)

With respect to the NXP MPU configuration that is part of this specific patch we are not targeting the scenarios that you are describing hence in my previous comment I was suggesting that at this stage without specific requirements the way to use mcuboot like applications is overriding CONFIG_HAS_SYSMPU in the application configuration setting it to n.

The support of the scenarios you described will be part of a future patch once the scope and the usecases are formalized.

dbkinder commented on arch/arm/core/cortex_m/mpu/arm_mpu.c:
```diff
@@ -10,9 +10,45 @@
+
+/**
+ * The attributes referenced in this function are described at:
+ * https://goo.gl/hMry3r
```
I'd recommend not using shortened urls, but the full url is longer than allowed by our coding standards:
http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dui0553a/Cihegaib.html
sigh.  I guess we're stuck with the shortened url :(

pfalcon replied:
Let's update coding standards away from 1970ies VT100 terminal with 80 chars per line...

fvincenzo replied:
Yes I think so. I used google at least it is a reliable service.

dbkinder reviewed (dismissed):
+1 documentation

mbolivar commented:
> Said that the patch you are referring to with your changes zephyrproject-rtos/zephyr@5951e45 has been in RFC for nearly 2 months and I was hoping to have captured all the requirements in terms of exceptions but it seems something it is still missing.

Yeah, I think we should have paid more attention to your efforts when they were at the RFC stage; hopefully it is better late than never.

> without specific requirements the way to use mcuboot like applications is overriding CONFIG_HAS_SYSMPU in the application configuration setting it to n.

I hope we can work together to try to capture some requirements here that will allow the MPU to be turned on even for mcuboot.

Let me in particular point out this patch that I had pasted earlier:

https://github.com/Linaro/zephyr/commit/14af77b21ce54e10a0c62139c988271f48776b23

We use this both in our mcuboot builds and for an application (https://github.com/linaro/zephyr-fota-hawkbit) which writes flash to download firmware over the air updates for mcuboot to use.

In current Zephyr, of course, this is the same as giving all code RW access to flash, since all code is privileged. But the intent is to try to allow privileged components of applications to read/write flash if they want to, without affecting the defaults that flash is read-only. This allows you to leave the MPU on, catching other bugs.

With the MPU on, we've so far seen:

- network stack issues in Zephyr (caught in the application)
- bad address calculations in mcuboot
- even bugs in the MPU configuration itself (caught in the application, addressed as https://github.com/Linaro/zephyr/commit/0b2a39cedf5ab1ed4d90efa361b14de89f347e06)

None of these would have been found if we had to turn the MPU off just because we wanted to write flash.

fvincenzo commented:
Ok, agreed, I am going to add the requested features to both NXP and ARM MPUs based on the requirements formalized via e-mail.

Once I pushed the the changed, I would like to ask to you to test them in your setup and let me know if the re is still something outstanding. :smile:

Commit:
samples: mpu_test: Add controller specific write

This patch add a controller specific write function and fixes a
documentation error in the mpu_test README file.

Signed-off-by: Vincenzo Frascino <vincenzo.frascino@linaro.org>

erwango commented on samples/mpu_test/src/main.c:
```diff
@@ -33,18 +34,78 @@ static int shell_cmd_read(int argc, char *argv[])
+	flash_dev = device_get_binding(CONFIG_SOC_FLASH_MCUX_DEV_NAME);
+
+	/* 16K reserved to the application */
+	u32_t offset = FLASH_MEM + 0x20000;
```
Is that 16K or 128k?

fvincenzo replied:
Should be 16 here. Thanks for catching this.

mbolivar commented:
Hi @fvincenzo; thanks for the update. I've tested this series on an mcuboot build for FRDM-K64F with CONFIG_MPU_ALLOW_FLASH_WRITE=y and verified that the bootloader is able to write to flash without causing an MPU violation.

fvincenzo commented:
@mbolivar Happy to hear that! I hope this feature helps you guys to find all the bugs and make Zephyr solid rock! 😄

mike-scott reviewed (commented):
Hi @fvincenzo,
Do you think it's possible to keep my authorship on the "Add Allow write on Flash" patch?  I know you changed the title and description a bit (and moved the Kconfig portion out into another patch), but essentially it's the same as: https://github.com/Linaro/zephyr/commit/14af77b21ce54e10a0c62139c988271f48776b23

fvincenzo commented:
@mike-scott The reason why I have my authorship in the patch it is because I did not use any of your work directly (means I did not apply your patch on my tree). I implemented the patch based on the requirements coming from the e-mail thread.
Since you did some pre-existing work I think it is correct to have your sign-off-by in there which I will add tomorrow once in office. Is that ok for you?

fvincenzo commented:
@mike-scott As promised I added your signed-off-by.

agross-linaro commented:
@mike-scott @fvincenzo 

Looking at this code, there isn't any different way to add security options for this single region.  There is nothing novel about the implementation.  As such, a signed-off-by should be fine.

mike-scott commented:
@fvincenzo thanks!
@agross-linaro : agreed :)

galak commented on arch/arm/core/cortex_m/mpu/Kconfig:
```diff
@@ -13,3 +13,18 @@ config ARM_MPU
+	help
+	  MCU has NXP MPU
+
+config MPU_ALLOW_FLASH_WRITE
```
we should separate this out into a different patch, since I don't see why we need a config MPU_ALLOW_FLASH_WRITE, can't we just do this if CONFIG_FLASH is enabled in the code?  Since that would imply you have the ability to write to the flash

mike-scott replied:
CONFIG_FLASH doesn't imply that you want to write to flash.  It needs to be an additional option.

galak replied:
Why would you enable FLASH drivers unless you wanted to write to flash?

mike-scott replied:
Hrm.. now that I look, there isn't in a use-case in Zephyr for enabling CONFIG_FLASH that isn't write/erase centric.   Prior to adding a CONFIG_* option to adjust MPU settings, I was thinking of a dynamic MPU hook which could be triggered via flash_write_protection_set() calls.  But, I'm not sure if we can do that in a generic way across MPU implementations.  But the side effect of that method would be no additional CONFIG options.

galak commented on include/arch/arm/cortex_m/mpu/nxp_mpu.h:
```diff
@@ -0,0 +1,72 @@
+#define MPU_REGION_SU    ((3 << 3) | (3 << 9) | (3 << 15) | (3 << 21))
+
+/* Some helper defines for common regions */
+#if defined(CONFIG_MPU_ALLOW_FLASH_WRITE)
```
why not #if defined(CONFIG_FLASH)

fvincenzo replied:
Agreed we can change this.

galak reviewed (changes requested):
I think we can just use CONFIG_FLASH and remove CONFIG_MPU_ALLOW_FLASH_WRITE

mbolivar commented:
One reason is to read from a flash device that isn't memory mapped

galak commented:
Ok, than you can use the flash driver cfg itself, CONFIG_SOC_FLASH_MCUX

fvincenzo commented:
In my opinion the configuration option should not be specific to the SOC or to the Flash Driver and the reason is because the application should be able to configure the behavior in the same way independently from the MCU is running on (I do not think we want to have a prj.conf for each MCU we are supporting in mcuboot). This is the reason why at the beginning I defined  CONFIG_MPU_ALLOW_FLASH_WRITE and agreed with changing it in CONFIG_FLASH (which should enable automatically the specific flash).

If CONFIG_FLASH does not cover all the scenarios, I can revert back to the original config option. Let me know what do you think.

MaureenHelm commented on samples/mpu_test/prj_allow_flash_write_mcux.conf:
```diff
@@ -0,0 +1,3 @@
+CONFIG_CONSOLE_SHELL=y
+CONFIG_FLASH=y
+CONFIG_SOC_FLASH_MCUX=y
```
You don't need to set CONFIG_SOC_FLASH_MCUX=y explicitly. The Kinetis SoC defconfigs enable it if CONFIG_FLASH=y

fvincenzo replied:
Thanks I will remove it.

MaureenHelm commented on samples/mpu_test/README.rst:
```diff
@@ -46,6 +46,13 @@ single thread: :file:`prj_single.conf`:
+   $ make BOARD=v2m_beetle CONF_FILE=prj_allow_flash_write_<x>.conf run
+
 Sample Output
 =============
```
I ran this sample on frdm_k64f and noticed the console output doesn't match what is documented in the README.rst.  Looks like perhaps the shell got added after the doc was written? Should probably fix in a follow up patch

fvincenzo replied:
Yes, right this sample changed a bit over time and I forgot to update the output.

MaureenHelm commented on samples/mpu_test/src/main.c:
```diff
@@ -33,18 +34,81 @@ static int shell_cmd_read(int argc, char *argv[])
+	flash_write_protection_set(flash_dev, false);
+
+	if (flash_write(flash_dev, offset, value,
+				2 * sizeof(u32_t)) != 0) {
```
Better to use sizeof(value)?

fvincenzo replied:
Ok, I can change it.

mbolivar commented:
> the application should be able to configure the behavior in the same way independently from the MCU is running on (I do not think we want to have a prj.conf for each MCU we are supporting in mcuboot)

From an mcuboot contributor perspective, I think this is spot on.

Note also that all Zephyr applications which are chain-loaded by mcuboot have to use their flash drivers to update some state shared with mcuboot, to e.g. signal that they booted correctly and don't need to be rolled back to an old version.

So if you make this SOC flash driver-specific, each of those applications will have to either disable the MPU or maintain a config file for every board they support, which doesn't scale. With the patch as-is, however, you can let the applications express their intent "I want to use the MPU, but I need to write my flash" in just one file.

fvincenzo commented:
@mbolivar From what you are saying seems that your preference is CONFIG_MPU_ALLOW_FLASH_WRITE. Am I correct?

MaureenHelm reviewed (commented):
> @mbolivar From what you are saying seems that your preference is CONFIG_MPU_ALLOW_FLASH_WRITE. Am I correct?

I think so, and I think I agree

andrewboie commented on samples/mpu_stack_guard_test/testcase.ini:
```diff
@@ -0,0 +1,10 @@
+tags = apps
+
+[test_stack_guard]
+build_only = true
```
why is this a build_only test?
There are ways of testing whether the system correctly caught an exception.
See tests/kernel/fatal for an example.
I'd like to see this moved to tests/ and verify correctness at runtime.

fvincenzo replied:
Thanks for pointing this out. I will adapt my code accordingly.

tyrantlizard77 reviewed (dismissed):
Is it possible to introduce an MPU configuration for qemu_cortex_m3 so that we can test some of these MPU features with sanitycheck?

galak commented:
I don't see why you end up with per MCU proj.conf files.  For a given SoC you will end up in the .config have both CONFIG_FLASH & CONFIG_WHATEVER_THE_DRIVER_IS enabled.  All that I'm saying is in nxp_mpu.h we'd use the CONFIG_SOC_FLASH_MCUX.

mike-scott commented:
@galak I think as @mbolivar pointed out reading from external (or non-memory mapped) flash would require enabling CONFIG_FLASH.  And that use-case doesn't equate to wanting to write to flash.

So, I agree with the others that a separate CONFIG to allow MPU access for flash writes is the best option for the moment.

mbolivar commented:
> I don't see why you end up with per MCU proj.conf files. For a given SoC you will end up in the .config have both CONFIG_FLASH & CONFIG_WHATEVER_THE_DRIVER_IS enabled. All that I'm saying is in nxp_mpu.h we'd use the CONFIG_SOC_FLASH_MCUX.

Hm, I must have misunderstood your proposal. Let me explain myself better.

To be clear, I do think CONFIG_MPU_ALLOW_FLASH_WRITE is worth having.

Here are the cases I can see:

1. User does not want to use the MPU, and may or may not want the flash driver
2. User wants to use the MPU, but does not want to write flash.
2.1. Sub-case where user wants the SoC flash driver, but to read flash only (I know that it's memory-mapped in this case and you could just read with load instructions, but it's better to use the driver.)
2.2. Sub-case where user does not want the SoC flash driver at all.
3. User wants MPU, and does want to write flash. In this case, the user wants the SoC flash driver, for reading and writing.

Here's how they could be supported in the presence of MPU_ALLOW_FLASH_WRITE.

1. MPU config = n, MPU_ALLOW_FLASH_WRITE isn't meaningful, SoC flash driver config = don't care
2.1. MPU config = y, MPU_ALLOW_FLASH_WRITE = n, SoC flash driver config = y
2.2 MPU config = y, MPU_ALLOW_FLASH_WRITE = n, SoC flash driver config = n
3. MPU config = y, MPU_ALLOW_FLASH_WRITE = y, SoC flash driver config = y

Can you explain how you'd support these without MPU_ALLOW_FLASH_WRITE?

galak commented:
I don't see why anyone would realistically use the flash drive to read.  Why is it better to use the driver, its just doing load/store for flashes that are memory mapped.  So what's better about this?

Just enumerating all cases of MPU enabled/disabled + MPU_ALLOW_FLASH_WRITE enabled/disable doesn't really justify the need.  3 or the 4 cases can be handled as I suggested.  The only case is for some reason of having the MPU enabled but not wanting to allow flash writing but enabling the driver, this seems broken to me as we now cause MPU errors for what is a reasonable API to call.  If you enable the FLASH I think its a pretty reasonable expectation that you should be able to write to it.

galak commented:
@mike-scott I agree that CONFIG_FLASH isn't sufficient, thus suggesting using the driver CONFIG option in the MPU code instead.

Additionally, even if we introduce CONFIG_MPU_ALLOW_FLASH_WRITE, I don't think it should be user selectable.

galak commented:
@andrewboie we've looked at MPU support for QEMU, problem is MMU support on QEMU which is the obvious thing to base this on assumes a 4k page size (or at least that the smallest is my understanding).  So trying to support an MPU would require huge effort in QEMU.

andrewboie commented:
@galak ah, makes sense. So qemu-system-arm only emulates the MMU and not any of the various MPUs out there. Bummer!

mbolivar commented:
> I don't see why anyone would realistically use the flash drive to read. Why is it better to use the driver, its just doing load/store for flashes that are memory mapped. So what's better about this?

If that is true, why is there a flash_read() function at all?

Having a flash driver lets applications read from flash without having to duplicate the code for where the base address is, etc. The results are more self-documenting, for one, and are also useful for applications which may in some configurations want to only read flash and in others to read and write.

If you want an example of such an application: mcuboot itself. There was a lot of feedback at the last Connect that expressed discomfort at how much work it was doing (A/B swapping being safe to do in the face of random power failures was the main complaint). Since then, we've been looking seriously at an mcuboot configuration that only reads flash, and picks a 'most recent' or 'best' application to boot.

If we can have separate CONF_FILEs that let us switch which boot style (A/B swap with read/write vs. boot only) and simultaneously change the MPU configuration (MPU+write flash vs. MPU+never write flash), then the same code can be reused to read image headers, run checksums, etc.

If that's not possible, then the "read-only" mcuboot has to basically include its own copy of a generic flash driver. If any hardware ever comes along that has quirky flash reads, then it'll have to copy those quirks as well.

So I think there's already a good reason to have this, and for it to be user selectable.

fvincenzo commented:
> I don't see why you end up with per MCU proj.conf files. For a given SoC you will end up in the .config have both CONFIG_FLASH & CONFIG_WHATEVER_THE_DRIVER_IS enabled.

At this stage of the Zephyr development this seems to be true only for the nordic_nrf5, the kinetis family and the arduino 101. Is that correct?

mbolivar commented:
@galak what if CONFIG_MPU_ALLOW_FLASH_WRITE is default y if CONFIG_FLASH?

That should keep the default sane for users that load the flash driver, but still allow users to test that their applications which read flash aren't erasing or writing it if they don't want them to.

galak commented:
So my feeling is to make progress we remove CONFIG_MPU_ALLOW_FLASH_WRITE for right now.  Than we can introduce a new PR which uses CONFIG_MPU_ALLOW_FLASH_WRITE once this PR gets merged.

MaureenHelm commented:
> both CONFIG_FLASH & CONFIG_WHATEVER_THE_DRIVER_IS enabled. All that I'm saying is in nxp_mpu.h we'd use the CONFIG_SOC_FLASH_MCUX.

While I'm not opposed to CONFIG_MPU_ALLOW_FLASH_WRITE, this works too.

The k64 related changes look good.

mbolivar commented:
No issues from me; we can carry that patch out of tree.
